src/build_vehicle_gui.cpp

```diff
@@ -269,20 +269,32 @@ static bool EnginePowerVsRunningCostSorter(const EngineID &a, const EngineID &b)
 {
 	const Engine *e_a = Engine::Get(a);
 	const Engine *e_b = Engine::Get(b);
-
-	/* Here we are using a few tricks to get the right sort.
-	 * We want power/running cost, but since we usually got higher running cost than power and we store the result in an int,
-	 * we will actually calculate cunning cost/power (to make it more than 1).
-	 * Because of this, the return value have to be reversed as well and we return b - a instead of a - b.
-	 * Another thing is that both power and running costs should be doubled for multiheaded engines.
-	 * Since it would be multiplying with 2 in both numerator and denominator, it will even themselves out and we skip checking for multiheaded. */
-	Money va = (e_a->GetRunningCost()) / max(1U, (uint)e_a->GetPower());
-	Money vb = (e_b->GetRunningCost()) / max(1U, (uint)e_b->GetPower());
-	int r = ClampToI32(vb - va);
-
-	/* Use EngineID to sort instead since we want consistent sorting */
-	if (r == 0) return EngineNumberSorter(a, b);
-	return _engine_sort_direction ? r > 0 : r < 0;
+	uint p_a = e_a->GetPower();
+	uint p_b = e_b->GetPower();
+	Money r_a = e_a->GetRunningCost();
+	Money r_b = e_b->GetRunningCost();
+	/* Check if running cost is zero in one or both engines.
+	   If only one of them is zero then that one is higher
+	   else if both are then compare powers.
+	*/
+	if (r_a == 0) {
+		if (r_b == 0) {
+			/* If it is ambiguous which to return go with their ID */
+			if (p_a == p_b) return EngineNumberSorter(a,b);
+			return _engine_sort_direction != (p_a < p_b);
+		}
+		return !_engine_sort_direction;
+	}
+	if (r_b == 0 ) return _engine_sort_direction;
+
+	/* Double for more precision when comparing close values */
+	/* This shouldn't have any major effects in performance nor in keeping
+	   the game in sync between players since it's used in GUI only in client side */
+	double va = (double)p_a / (double)r_a;
+	double vb = (double)p_b / (double)r_b;
+	/* Use EngineID to sort if both have zero power instead since we want consistent sorting */
```

Comment:
It's not just zero power, va == vb checks for equal power/cost ratio in some (but not all) cases.

Reply:
Yes. I will change that comment to something like the above comment.